system-contracts/test/L1Messenger.spec.ts

```diff
@@ -0,0 +1,517 @@
+import { ethers, network } from "hardhat";
+import type { L1Messenger } from "../typechain";
+import { L1MessengerFactory } from "../typechain";
+import { prepareEnvironment, setResult } from "./shared/mocks";
+import type { StateDiff } from "./shared/utils";
+import { compressStateDiffs, deployContractOnAddress, encodeStateDiffs, getCode, getWallets } from "./shared/utils";
+import { utils } from "zksync-web3";
+import type { Wallet } from "zksync-web3";
+import {
+  TEST_KNOWN_CODE_STORAGE_CONTRACT_ADDRESS,
+  TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS,
+  TEST_BOOTLOADER_FORMAL_ADDRESS,
+  TWO_IN_256,
+} from "./shared/constants";
+import { expect } from "chai";
+import { BigNumber } from "ethers";
+import { randomBytes } from "crypto";
+
+describe("L1Messenger tests", () => {
+  let l1Messenger: L1Messenger;
+  let wallet: Wallet;
+  let l1MessengerAccount: ethers.Signer;
+  let knownCodeStorageAccount: ethers.Signer;
+  let bootloaderAccount: ethers.Signer;
+  let stateDiffsSetupData: StateDiffSetupData;
+  let logData: LogData;
+  let logs: string[];
+  let bytecodeData: BytecodeData;
+  let emulator: L1MessengerPubdataEmulator;
+
+  before(async () => {
+    await prepareEnvironment();
+    wallet = getWallets()[0];
+    await deployContractOnAddress(TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS, "L1Messenger");
+    l1Messenger = L1MessengerFactory.connect(TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS, wallet);
+    l1MessengerAccount = await ethers.getImpersonatedSigner(TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS);
+    knownCodeStorageAccount = await ethers.getImpersonatedSigner(TEST_KNOWN_CODE_STORAGE_CONTRACT_ADDRESS);
+    bootloaderAccount = await ethers.getImpersonatedSigner(TEST_BOOTLOADER_FORMAL_ADDRESS);
+    // setup
+    stateDiffsSetupData = await setupStateDiffs();
+    logData = setupLogData();
+    logs = createLogs(l1MessengerAccount, l1Messenger, logData);
+    bytecodeData = await setupBytecodeData(l1Messenger.address);
+    await setResult("SystemContext", "txNumberInBlock", [], {
+      failure: false,
+      returnData: ethers.utils.defaultAbiCoder.encode(["uint16"], [1]),
+    });
+    emulator = new L1MessengerPubdataEmulator();
+  });
+
+  // this part is necessary to clean the state of L1Messenger contract
+  after(async () => {
+    emulator.addLog(logs[0]);
```

Comment:
Could you please elaborate on why do we need so many operations here? 

I understand that we likely need to clear the storage, but a single `publishPubdataAndClearState` should be enough, no? I.e. why do we push new items into the L1Messenger

Reply:
This is caused by a change in the state of L1Messenger from positive test cases of sendL2ToL1Log, sendToL1, and requestBytecodeL1Publication. One possible solution would be to add this data during the positive test cases mentioned above, but I decided to completely reset the state in the 'after' block, instead of using `publishPubdataAndClearState ` in the last test case.

Reply:
fixed